### PR TITLE
Docker Compose based developer workflow

### PR DESCRIPTION
**Pull request by rosskarchner**
Provides a docker-compose based developer workflow.

Check out the instructions for [Installation](https://github.com/cfpb/cfgov-refresh/blob/10minutes/docs/installation.md#docker-compose-installation) and [Usage](https://github.com/cfpb/cfgov-refresh/blob/10minutes/docs/usage.md#usage-docker)

## Additions

- configuration and documentation for Docker and Docker Compose
- a script for creating an initial Docker Machine VM.
- Scripts that ease interacting with the Docker/Compose environment

## Testing

- Verify that checking out this branch does not break your existing workflow. Everything you do on a normal basis should continue to work. If it does, please report it here.
- Try out the installation instructions.
- Read the usage instructions.

## Todos

- This setup presumes that you still have Node, gulp, etc installed on the mac side. I think with further work, we can eliminate that dependency.
- We need a mechanism for SSL server.

## Checklist

* [x] PR has an informative and human-readable title
* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [development playbook](https://github.com/cfpb/development)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated
* [ ] Reviewers requested with the [Assignee tool](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) :arrow_right:

### CODE REVIEW

**Comment by rosskarchner** [comment]
Glitches that surfaced during the live demo:
- `dbshell` requires you to have previously done `eval $(docker-machine env cfgov)`
- the awk command in refresh-data.sh doesn't (always?) capture the container name

**Comment by marcesher** [comment]
@rosskarchner if I follow the install instructions to the letter, I see two strange things:

1. `source setup.sh` runs and then closes the terminal window I've run it in. This is in iTerm2 for what it's worth. I can consistently reproduce this behavior.

1. if I `source start-services.sh` I get this: `Get https://registry-1.docker.io/v2/cfgov/pdfreactor/manifests/latest: unauthorized: incorrect username or password`

1. Presumably because of ☝️failures, when I run `runserver` I get this: 

```
Starting the Django server on port 8000 ...
Traceback (most recent call last):
  File "cfgov/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ImportError: No module named django.core.management
```

**Comment by marcesher** [comment]
ok, after `workon`, next error:

```
  File "/Users/esherm/dev/.virtualenvs/cfgov-refresh/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 27, in <module>
    raise ImproperlyConfigured("Error loading MySQLdb module: %s" % e)
django.core.exceptions.ImproperlyConfigured: Error loading MySQLdb module: dlopen(/Users/esherm/dev/.virtualenvs/cfgov-refresh/lib/python2.7/site-packages/_mysql.so, 2): Library not loaded: /usr/local/homebrew/opt/openssl/lib/libssl.1.0.0.dylib
  Referenced from: /Users/esherm/dev/.virtualenvs/cfgov-refresh/lib/python2.7/site-packages/_mysql.so
```

**Comment by rosskarchner** [comment]
Thanks, Mark! That might be two different problems. 

For the first thing: this *might* be an artifact of your previous experiments with Docker -- an expired authentication to docker hub. Try `docker logout`

For the second thing, I think the quickstart instructions might be missing a `workon cfgov-refresh` before the runserver step.

**Comment by rosskarchner** [comment]
That's interesting-- can you `rmvirtualenv cfgov-refresh` AND `rm -rf ~/Library/Caches/pip` and try again?

If that doesn't work, maybe we need to add openssl to homebrew-deps.sh

**Comment by rosskarchner** [comment]
In fact, reverse that: try `brew install openssl` before trying the other thing.

**Comment by marcesher** [comment]
openssl already installed.

I tried removing the virtualenv and clearing the pip cache. then sourcing env and setup, and same behavior.

Oh, that docker logout worked, by the way. thanks!

**Comment by rosskarchner** [comment]
OK, thanks for the report-- it seems like the pre-built MySQL wheel isn't as bulletproof as I thought, which is annoying.

**Comment by rosskarchner** [comment]
Is there a /usr/local/homebrew/opt/openssl/lib/libssl.1.0.0.dylib on your system?

**Comment by rosskarchner** [comment]
This seems to suggest that maybe reinstalling openssl is a fix: https://stackoverflow.com/a/20946142

**Comment by marcesher** [comment]
I'll give that a shot, Ross. Thanks.

**Comment by marcesher** [comment]
@rosskarchner interesting... that libssl.dylib is looking in /usr/local/homebrew/opt but I recently had to wipe out my homebrew and start over, and I bet there are remnants on my mac that are still trying to point things to there. I have not been able to figure out why certain things still try to point to /usr/local/homebrew instead of the homebrew in my home directory. That file definitely lives in the homebrew in my home dir.

**Comment by rosskarchner** [comment]
hmm, maybe the built _mysql.so has a hardcoded path to /usr/local/homebrew

**Comment by marcesher** [comment]
ok @rosskarchner, more fun. I copied openssl into /usr/local/homebrew and that got me past the problem above. Now, when I run `runserver` it's telling me it can't connect to mysql. But... I see it running. Also, if I try hitting both elasticsearch and pdfreactor in a browser (http://localhost:9200 and http://localhost:9423), I get "can't connect".

```
(cfgov-refresh) serenity@<cfgov-refresh-ten> (10minutes)🐶  docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                            NAMES
d0a9866f917f        elasticsearch:2.3   "/docker-entrypoin..."   2 hours ago         Up About an hour    0.0.0.0:9200->9200/tcp, 0.0.0.0:9300->9300/tcp   cfgovrefreshten_elasticsearch_1
22b883845425        cfgov/pdfreactor    "/docker-entrypoin..."   2 hours ago         Up About an hour    9200/tcp, 9300/tcp, 0.0.0.0:9423->9423/tcp       cfgovrefreshten_pdfreactor_1
79c3e20930cc        mysql:5.7           "docker-entrypoint..."   2 hours ago         Up About an hour    0.0.0.0:3306->3306/tcp                           cfgovrefreshten_mysql_1
```

File: README.md [hunk 1]
```diff
@@ -18,11 +18,11 @@ Full [installation](https://cfpb.github.io/cfgov-refresh/installation/)
...
-and our
-[runserver.sh script](https://github.com/cfpb/cfgov-refresh/blob/master/runserver.sh#L12).
+Ensure that Python, Node, Docker, Docker Machine, Docker Compose, and imagemagick are installed. If you are using a mac with homebrew, you can run:
+
+`sh homebrew-deps.sh`
+
+Then:
 
 ```
 git clone git@github.com:cfpb/cfgov-refresh.git
```
Comment (willbarton) [inline README.md]: Can't wait to try this, but just to point out, this currently has users running `homebrew-deps.sh` before checking out cfgov-refresh ;)

**Comment by rosskarchner** [comment]
OK Marc-- Can you either:

`docker-machine rm cfgov` and `./start-services.sh` again, OR

Run these commands by hand: https://github.com/cfpb/cfgov-refresh/blob/10minutes/docker-environment.sh#L16-L19

**Comment by rosskarchner** [comment]
I'm working on some changes here:

- Instead of a pre-built MySQLdb wheel, we'll use pymysql for local development
- Since PDFReactor is on the way out, I've removed it from our docker-compose setup

**Comment by rosskarchner** [comment]
The documentation above should no longer be considered correct-- stay tuned!

**Comment by rosskarchner** [comment]
@willbarton I just took off the 'hold' label ;)

**Comment by rosskarchner** [comment]
One last pre-merge thank you to @wpears!

**Pull request merged**